### PR TITLE
Emit log on Runtime Code change.

### PR DESCRIPTION
**Pull request by tomusdrw**
Related: #5047 

Currently, we only emit an event in case the runtime code changes. However as described in #5047, for light clients it's difficult to track the code changes, cause reading events is not really feasible.

This PR introduces a `RuntimeUpdated` digest item, which is triggered on every code update (via `frame_system` dispatchables) or on `heap_pages` change (also via `frame_system` function, not direct state write).

This will allow the light clients to easily update the current code, based just on reading the digest items in the header.

### Upgrade considerations

Since `DigestItem` is used in the client code as well (i.e. we decode the header digest), the nodes MUST be updated before rolling out the runtime upgrade.

### CODE REVIEW

File: frame/system/src/lib.rs
Comment (thiolliere) [inline frame/system/src/lib.rs:1131]: needs to run cargo fmt.

File: primitives/runtime/src/generic/digest.rs
Comment (thiolliere) [inline primitives/runtime/src/generic/digest.rs:128]: I personally wouldn't mind the more accurate name `RuntimeOrHeapPagesUpdated`

**Review by thiolliere — APPROVED** [review]
code looks good to me

File: frame/system/src/lib.rs
Comment (bkchr) [inline frame/system/src/lib.rs:1074]: This will not work for Parachains.

Parachains have a special `set_code` implementation that will not update `:code` here directly. It will first queue the update and update it when it is allowed by the relay chain.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: frame/system/src/lib.rs
Comment (tomusdrw) [reply frame/system/src/lib.rs:1074]: Where can we find the queue implementation? Any chance the actual storage write can go through some function that would emit the log?

Comment (bkchr) [reply frame/system/src/lib.rs]: Yeah we can do this.

https://github.com/paritytech/cumulus/blob/master/pallets/parachain-system/src/lib.rs#L943

Comment (tomusdrw) [reply frame/system/src/lib.rs]: I've exposed a helper function to update the storage and emit event + log. The idea is that this function will be used by parachain code implementation and is used by the default implementation of `OnSetCode` for unit type.

**Review by bkchr — APPROVED** [review]

**Comment by tomusdrw** [comment]
bot merge

File: primitives/runtime/src/generic/digest.rs
Comment (pepyakin) [inline primitives/runtime/src/generic/digest.rs:196]: Right now the story about heap pages is in flux. AFAIU, there is an interest to get rid of heap pages and or change the semantics at least.

In that case, maybe it would be better to split those digests in two: runtime code updated and heap pages updated? Then, if needed we deprecate the heap pages. The alternative would be to leave it as is and then change the semantics of `RuntimeCodeOrHeapPagesUpdated` which seems a bit messier.

**Review by pepyakin — COMMENTED** [review]
Sorry I am late to party, just discovered this PR.

File: primitives/runtime/src/generic/digest.rs
Comment (tomaka) [reply primitives/runtime/src/generic/digest.rs:196]: I preferred `RuntimeUpdated`, as it simply means "you might want to check whatever storage item concerns the runtime". The ambiguity is advantageous here.

Comment (tomusdrw) [reply primitives/runtime/src/generic/digest.rs:196]: @thiolliere @bkchr are you guys okay to revert back to `RuntimeUpdated` so that we can be a bit more flexible regarding what that actually means? My preference would be to have a single enum variant, especially if the heap pages one might get deprecated in the future.

Comment (thiolliere) [reply primitives/runtime/src/generic/digest.rs:196]: I also think that once we remove HeapPages stuff, RuntimeCodeOrHeapPagesUpdated always mean RuntimeCodeUpdated.
So the transition shouldn't be that messy. Anyway it is ok to me.

**Comment by cheme** [comment]
quick question before this is merged, (not sure it is a good idea): should we check at https://github.com/paritytech/substrate/blob/864d096214cbde236f1ced7fdf4017ac22f5e5f5/client/db/src/lib.rs#L2009 or somewhere else after block import processing that when the digest is present in block import header, we indeed have an item for :code in the storage change payload (both are field of block import).
This way we ensure the digest is not added by the block builder without a storage update (same test could be added to cumulus).

**Comment by tomusdrw** [comment]
> This way we ensure the digest is not added by the block builder without a storage update (same test could be added to cumulus).

Do we do that for any other digest item? Note it's not an inherent (i.e. not block author's opinion), but rather a digest item and afair the only way to add them is through the runtime execution? That means that a block with an "extra" digest item would be considered invalid, because it wouldn't be generated by the runtime during `execute_block`.
Or am I missing something?

**Comment by cheme** [comment]
I am not sure, but on a recent dropped branch I did manage some digest directly with the block builder after https://github.com/paritytech/substrate/blob/864d096214cbde236f1ced7fdf4017ac22f5e5f5/client/block-builder/src/lib.rs#L226, and a check on block import.
Yet the check on block import would have need a cumulus implementation to.

**Comment by cheme** [comment]
https://github.com/paritytech/substrate/blob/864d096214cbde236f1ced7fdf4017ac22f5e5f5/frame/executive/src/lib.rs#L459 should avoid issue indeed.

**Comment by tomaka** [comment]
One short coming with this PR's design is that, in case of a fork at the block height that modifies the runtime, all forks will have that log item but a light client will not be able to know whether the forks did all update to the same runtime or to different ones. Including a hash of the code in the digest item would solve this, but would make the design more restrictive overall. I don't think it's a big deal to leave that unaddressed.

**Comment by tomusdrw** [comment]
> Including a hash of the code in the digest item would solve this,

That's actually something I wanted to add initially, but I don't think there is an efficient way to get the hash of the code without actually computing it on-chain. (i.e. no way to read storage item hash after write afaict).

Also this complicates a bit the "heap pages" problem, since in such case the code hash stays the same. Unless we make the code hash optional or hash all of the runtime environment parameters (which might once again be a bit expensive).

File: primitives/runtime/src/generic/digest.rs
Comment (tomusdrw) [reply primitives/runtime/src/generic/digest.rs]: Reverted back to `RuntimeUpdated`.

**Comment by tomaka** [comment]
Shall we merge this? Is there any reason left to hold off this PR?

**Comment by bkchr** [comment]
None that I'm aware of.

File: primitives/runtime/src/generic/digest.rs
Comment (kianenigma) [reply primitives/runtime/src/generic/digest.rs:128]: `RuntimeEnvironmentUpdated`

**Comment by tomaka** [comment]
bot merge

**Comment by apopiak** [comment]
Late to the party: What about including the pre- and post-runtime version in the digest?

**Comment by tomusdrw** [comment]
I was considering including code hash in the digest ([see here](https://github.com/paritytech/substrate/pull/9580#issuecomment-909209398)), but yeah the heap pages complicate that a bit - would probably be easier to have two separate digest items.
Note that the event is only an indication for the light clients, and without any extra data, the light client can query any storage items when indicated to figure out what exactly has changed - I think it's way more flexible than attempting to add any extra data directly to the digest item.

**Comment by apopiak** [comment]
Yeah, unfortunate that heap pages and code were lumped together.

**Pull request merged**
